File: questModule/src/main/java/ru/nekit/android/qls/quest/answer/shared/AbstractGroupWeightComparisonQuestAnswerChecker.java
package ru.nekit.android.qls.quest.answer.shared;

import android.support.annotation.NonNull;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import ru.nekit.android.qls.quest.base.IGroupWeightComparisonQuest;
import ru.nekit.android.qls.quest.base.Quest;

public abstract class AbstractGroupWeightComparisonQuestAnswerChecker<T>
        implements IAnswerChecker<T> {

    protected abstract List<T> getGroupList(@NonNull Quest quest);

    @Override
    public boolean checkAlternativeInput(@NonNull Quest quest, @NonNull T answer) {
        IGroupWeightComparisonQuest inQuest = (IGroupWeightComparisonQuest) quest;
        HashMap<T, Integer> map = new HashMap<>();
        List<T> groupList = getGroupList(quest);
        final int length = groupList.size();
        int i = 0;
        for (; i < length; i++) {
            T key = groupList.get(i);
            Integer value = map.get(key);
            value = value == null ? 0 : value;
            map.put(key, value + 1);
        }
        boolean isMax = inQuest.getGroupComparisonType() == IGroupWeightComparisonQuest.MAX_GROUP_WEIGHT;
        int value = isMax ? 0 : Integer.MAX_VALUE;
        for (Map.Entry<T, Integer> entryItem : map.entrySet()) {
            int item = entryItem.getValue();
            value = isMax ? Math.max(value, item) : Math.min(value, item);
        }
        return map.get(answer) == value;
    }

    @Override
    public boolean checkStringInputFormat(@NonNull Quest quest, @NonNull String value) {
        return false;
    }

    @Override
    public boolean checkStringInput(@NonNull Quest quest, @NonNull String value) {
        return false;
    }

}

File: questDomain/src/main/java/ru/nekit/android/qls/domain/answerChecker/common/AbstractGroupWeightComparisonQuestAnswerChecker.kt
package ru.nekit.android.qls.domain.answerChecker.common

import ru.nekit.android.qls.domain.model.quest.Quest
import ru.nekit.android.qls.domain.model.resources.common.IGroupWeightComparisonQuest
import java.util.*

abstract class AbstractGroupWeightComparisonQuestAnswerChecker<in R : Quest> : IAnswerChecker<R> {

    protected abstract fun getGroupList(quest: R): List<Any>

    override fun checkAlternativeInput(quest: R, answer: Any): Boolean {
        val inQuest = quest as IGroupWeightComparisonQuest
        val map = HashMap<Any, Int>()
        val groupList = getGroupList(quest)
        val length = groupList.size
        var i = 0
        while (i < length) {
            val key = groupList[i]
            val value: Int = map[key] ?: 0
            map[key] = value + 1
            i++
        }
        val isMax = inQuest.groupComparisonType == IGroupWeightComparisonQuest.MAX_GROUP_WEIGHT
        var value = if (isMax) 0 else Integer.MAX_VALUE
        for ((_, item) in map) {
            value = if (isMax) Math.max(value, item) else Math.min(value, item)
        }
        return map[answer] == value
    }

    override fun checkStringInputFormat(quest: R, value: String): Boolean {
        return false
    }

    override fun checkStringInput(quest: R, value: String): Boolean {
        return false
    }

}